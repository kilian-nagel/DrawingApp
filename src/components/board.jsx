import React, { Component } from 'react';
import "../styles/board.css";

function Board() {
    return (
        <div className="board" id="board">

        </div>
    );
}

export default Board;