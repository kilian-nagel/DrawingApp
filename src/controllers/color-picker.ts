import { dblClick } from "@testing-library/user-event/dist/click";
import React from "react";

export let pencilColor:string = "abcdef";

export function handleColorPick(e: React.ChangeEvent<HTMLInputElement>){
    setPencilColor(e.currentTarget.value);
    setInputColor();
}

function setPencilColor(color:string){
    pencilColor = color;
}

function setInputColor(){
    console.log(pencilColor);
    (document.querySelector(".color-picker") as HTMLInputElement).value = pencilColor;
}