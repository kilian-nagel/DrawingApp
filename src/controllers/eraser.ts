import React from "react";


export function handleEraserClick(e:React.MouseEvent){

}