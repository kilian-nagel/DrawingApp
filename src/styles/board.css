
#board {
    flex-grow: 1;
    min-height: 800px;
    position: relative;
}

.pixel {
    position: absolute;
    width: 30px;
    height: 30px;
    border-radius: 16px;
}