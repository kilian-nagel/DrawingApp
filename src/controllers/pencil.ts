
import {changeActiveTool, parsePixel} from "./board";

export function handlePencilClick(e: React.MouseEvent){
    changeActiveTool(0);
    parsePixel(e);
}

export {};