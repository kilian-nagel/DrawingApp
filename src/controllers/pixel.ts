import { activeTool, board_pixels , board } from "./board";
import { pencilColor } from "./color-picker";
import { eraserColor } from "./eraser";

export function parsePixel(e: React.MouseEvent | MouseEvent){
    const pixelPosition = getPixelPosition(e);
    const element = createPixelElt(pixelPosition);
    addPixelEltToBoard(element);
    addPixelToBoardArray(pixelPosition);
}

export function createPixelElt(pixel_position:number[]){
    let elt = document.createElement("div") as HTMLElement;
    elt.classList.add("pixel");
    elt.classList.add("pixel-"+pixel_position[0]+"-"+pixel_position[1]);
    elt.style.background = `#${getCurrentColor()}`;
    elt.style.left = pixel_position[0]+"px";
    elt.style.top = pixel_position[1]+"px";
    return elt;
}

function changePixelColor(pixel_position:number[]){
    let elt = document.querySelector(".pixel-"+pixel_position[0]+"-"+pixel_position[1]) as HTMLElement
    elt.style.background = pencilColor;
}

function addPixelEltToBoard(elt:HTMLElement){
    board.appendChild(elt);
}

function addPixelToBoardArray(pixel_position:number[]){
    board_pixels[pixel_position[1]][pixel_position[0]]= pencilColor;
}

export function getPixelPosition(e: React.MouseEvent | MouseEvent):number[]{
    let pixel_pos_y = e.pageY-board.offsetTop;
    let pixel_pos_x = e.pageX;
    return [pixel_pos_x,pixel_pos_y];
}

function getCurrentColor():string{
    if(activeTool===0){return pencilColor;}
    if(activeTool===1){return eraserColor;}
    return "FFFFFF";
}