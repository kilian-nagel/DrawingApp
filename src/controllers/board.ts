import { pencilColor } from "./color-picker";
import { handlePencilClickEvent } from "./pencil";
import { eraserColor, handleEraserClickEvent } from "./eraser";
import React from "react";

export let activeTool = 0;
let board_pixels:string[][] = [];
let board:HTMLElement;

/* Click event redirection
=============== */

export function changeActiveTool(n:number){
    activeTool = n;
}

/* When a click event is fired on the board , 
the function determines wich tool has to be activated 
( pencil , eraser...) */
export function redirectClickEvent(event:React.MouseEvent,isMouseDown:boolean){
    if(activeTool===0){handlePencilClickEvent(event,isMouseDown);}
    else if(activeTool===1){handleEraserClickEvent(event,isMouseDown);}
}


/* Board stuff 
================ */

export function initBoardPixels(){
    board = document.getElementById("board") as HTMLElement;
    for(let i=0;i<board.clientHeight;i++){
        board_pixels.push([]);
        for(let y=0;y<board.clientWidth;y++){
            board_pixels[board_pixels.length-1].push("FFFFFF");
        }
    }
    
}

/* Pixel stuff
=============== */

export function parsePixel(e: React.MouseEvent | MouseEvent){
    const pixelPosition = getPixelPosition(e);
    createPixelElt(pixelPosition);
}

function getPixelPosition(e: React.MouseEvent | MouseEvent){
    let pixel_pos_y = e.pageY-board.offsetTop;
    let pixel_pos_x = e.pageX;
    return [pixel_pos_x,pixel_pos_y];
}

function createPixelElt(pixel_position:number[]){

    let background="FFFF";
    if(activeTool===0){background=pencilColor;}
    else if(activeTool===1){background=eraserColor}

    let elt = document.createElement("div") as HTMLElement;
    elt.classList.add("pixel-"+pixel_position[0]+"-"+pixel_position[1]);
    elt.style.background = `#${background}`;
    elt.style.position = "absolute";
    elt.style.left = pixel_position[0]+"px";
    elt.style.top = pixel_position[1]+"px";
    elt.style.width = "30px";
    elt.style.height = "30px";
    board.appendChild(elt);
    
    board_pixels[pixel_position[1]][pixel_position[0]]= pencilColor;
}

function changePixelColor(pixel_position:number[]){
    let elt = document.querySelector(".pixel-"+pixel_position[0]+"-"+pixel_position[1]) as HTMLElement
    elt.style.background = pencilColor;
}

function isPixelOnBoard(pixel_position:number[]){
    let pixel_element = document.querySelector(".pixel-"+pixel_position[0]+"-"+pixel_position[1]) as HTMLElement;
    return document.body.contains(pixel_element)
}
export {};