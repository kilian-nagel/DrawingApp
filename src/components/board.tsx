import React, { useEffect } from 'react';
import { initBoardPixels, redirectClickEvent } from '../controllers/board';
import { handlePencilClickEvent } from '../controllers/pencil';
import "../styles/board.css";

function Board() {
    let isInit = false;
    useEffect(()=>{
        if(!isInit){
            initBoardPixels();
            isInit = true;
        }
    },[])
    return (
        <div className="board" id="board" onMouseUp={(e)=>handlePencilClickEvent(e,false)} onMouseDown={(e)=>handlePencilClickEvent(e,true)}>

        </div>
    );
}

export default Board;