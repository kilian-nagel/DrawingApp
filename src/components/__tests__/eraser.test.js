import { render , screen , cleanup, fireEvent} from "@testing-library/react";
import '@testing-library/jest-dom';
import Drawbar from "../drawbar";
import { eraserColor } from "../../controllers/eraser";
import { board_pixels, changeActiveTool , initBoardPixels} from "../../controllers/board";
import Board from "../board";

afterEach(cleanup);

describe("eraser HTML Element",()=>{
    it("eraser is in the document",()=>{
        render(<Drawbar/>);
        const eraser = screen.getByTestId("eraser");
        expect(eraser).toBeInTheDocument();
    })
})

describe("eraser controller",()=>{
    it("eraser color initialized at #FFFFFF",()=>{
        render(<Drawbar/>);
        expect(eraserColor).toBe("FFFFFF");
    })
    it("on click eraser should erase",()=>{
        render(<Board/>);
        changeActiveTool(1);
        initBoardPixels();

        board_pixels[120][100]="AFAFAF";
        const board = screen.getByTestId("board");
        fireEvent.click(board,{clientX:100,clientY:120});
        expect(board_pixels[120][100]).toBe("FFFFFF");
    })
})