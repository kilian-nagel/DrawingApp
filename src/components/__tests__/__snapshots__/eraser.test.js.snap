// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`eraser HTML Element color picker renders correctly 1`] = `"<button aria-label=\\"sÃ©lectionner la gomme\\" data-testid=\\"eraser\\"><i class=\\"fa-solid fa-eraser fa-1x\\"></i></button>"`;
